File: Kernel Simulator/Shell/Shells/TextShell.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Threading
Imports KS.Misc.Editors.TextEdit
Imports KS.Shell.Prompts

Namespace Shell.Shells
    Public Class TextShell
        Inherits ShellExecutor
        Implements IShell

        Public Overrides ReadOnly Property ShellType As ShellType Implements IShell.ShellType
            Get
                Return ShellType.TextShell
            End Get
        End Property

        Public Overrides Property Bail As Boolean Implements IShell.Bail

        Public Overrides Sub InitializeShell(ParamArray ShellArgs() As Object) Implements IShell.InitializeShell
            'Get file path
            Dim FilePath As String = ""
            If ShellArgs.Length > 0 Then
                FilePath = ShellArgs(0)
            Else
                Write(DoTranslation("File not specified. Exiting shell..."), True, GetConsoleColor(ColTypes.Error))
                Bail = True
            End If

            'Actual shell logic
            While Not Bail
                Try
                    'Open file if not open
                    If TextEdit_FileStream Is Nothing Then
                        Wdbg(DebugLevel.W, "File not open yet. Trying to open {0}...", FilePath)
                        If Not TextEdit_OpenTextFile(FilePath) Then
                            Write(DoTranslation("Failed to open file. Exiting shell..."), True, GetConsoleColor(ColTypes.Error))
                            Bail = True
                            Exit While
                        End If
                        TextEdit_AutoSave.Start()
                    End If

                    'See UESHShell.vb for more info
                    SyncLock GetCancelSyncLock(ShellType)
                        'Prepare for prompt
                        If DefConsoleOut IsNot Nothing Then
                            Console.SetOut(DefConsoleOut)
                        End If
                        WriteShellPrompt(ShellType)

                        'Raise the event
                        KernelEventManager.RaiseTextShellInitialized()
                    End SyncLock

                    'Prompt for command
                    Dim WrittenCommand As String = ReadLine()
                    If Not (WrittenCommand = Nothing Or WrittenCommand?.StartsWithAnyOf({" ", "#"})) Then
                        KernelEventManager.RaiseTextPreExecuteCommand(WrittenCommand)
                        GetLine(WrittenCommand, False, "", ShellType.TextShell)
                        KernelEventManager.RaiseTextPostExecuteCommand(WrittenCommand)
                    End If
                Catch taex As ThreadInterruptedException
                    CancelRequested = False
                    Bail = True
                Catch ex As Exception
                    WStkTrc(ex)
                    Write(DoTranslation("There was an error in the shell.") + NewLine + "Error {0}: {1}", True, color:=GetConsoleColor(ColTypes.Error), ex.GetType.FullName, ex.Message)
                    Continue While
                End Try
            End While

            'Close file
            TextEdit_CloseTextFile()
            TextEdit_AutoSave.Stop()
        End Sub

    End Class
End Namespace

File: Kernel Simulator/Shell/Shells/TextShell.cs
using System;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System.Threading;
using KS.ConsoleBase.Colors;
using KS.ConsoleBase.Inputs;
using KS.Kernel;
using KS.Languages;
using KS.Misc.Editors.TextEdit;
using KS.Misc.Text;
using KS.Misc.Writers.ConsoleWriters;
using KS.Misc.Writers.DebugWriters;
using KS.Shell.Prompts;
using KS.Shell.ShellBase;
using KS.Shell.ShellBase.Shells;

namespace KS.Shell.Shells
{
	public class TextShell : ShellExecutor, IShell
	{

		public override ShellType ShellType
		{
			get
			{
				return ShellType.TextShell;
			}
		}

		public override bool Bail { get; set; }

		public override void InitializeShell(params object[] ShellArgs)
		{
			// Get file path
			string FilePath = "";
			if (ShellArgs.Length > 0)
			{
				FilePath = Convert.ToString(ShellArgs[0]);
			}
			else
			{
				TextWriterColor.Write(Translate.DoTranslation("File not specified. Exiting shell..."), true, KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.Error));
				Bail = true;
			}

			// Actual shell logic
			while (!Bail)
			{
				try
				{
					// Open file if not open
					if (TextEditShellCommon.TextEdit_FileStream is null)
					{
						DebugWriter.Wdbg(DebugLevel.W, "File not open yet. Trying to open {0}...", FilePath);
						if (!TextEditTools.TextEdit_OpenTextFile(FilePath))
						{
							TextWriterColor.Write(Translate.DoTranslation("Failed to open file. Exiting shell..."), true, KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.Error));
							Bail = true;
							break;
						}
						TextEditShellCommon.TextEdit_AutoSave.Start();
					}

					// See UESHShell.vb for more info
					lock (CancellationHandlers.GetCancelSyncLock(ShellType))
					{
						// Prepare for prompt
						if (Kernel.Kernel.DefConsoleOut is not null)
						{
							Console.SetOut(Kernel.Kernel.DefConsoleOut);
						}
						PromptPresetManager.WriteShellPrompt(ShellType);

						// Raise the event
						Kernel.Kernel.KernelEventManager.RaiseTextShellInitialized();
					}

					// Prompt for command
					string WrittenCommand = Input.ReadLine();
					if ((string.IsNullOrEmpty(WrittenCommand) | (WrittenCommand?.StartsWithAnyOf([" ", "#"]))) == false)
					{
						Kernel.Kernel.KernelEventManager.RaiseTextPreExecuteCommand(WrittenCommand);
						Shell.GetLine(WrittenCommand, false, "", ShellType.TextShell);
						Kernel.Kernel.KernelEventManager.RaiseTextPostExecuteCommand(WrittenCommand);
					}
				}
				catch (ThreadInterruptedException taex)
				{
					Flags.CancelRequested = false;
					Bail = true;
				}
				catch (Exception ex)
				{
					DebugWriter.WStkTrc(ex);
					TextWriterColor.Write(Translate.DoTranslation("There was an error in the shell.") + Kernel.Kernel.NewLine + "Error {0}: {1}", true, color: KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.Error), ex.GetType().FullName, ex.Message);
					continue;
				}
			}

			// Close file
			TextEditTools.TextEdit_CloseTextFile();
			TextEditShellCommon.TextEdit_AutoSave.Stop();
		}

	}
}